src/parser/mod.rs

```diff
@@ -10353,70 +10353,84 @@ impl<'a> Parser<'a> {
         }
     }
 
-    /// Parse a possibly qualified, possibly quoted identifier, optionally allowing for wildcards,
+    /// Parse a possibly qualified, possibly quoted identifier, e.g.
+    /// `foo` or `myschema."table"
+    ///
+    /// The `in_table_clause` parameter indicates whether the object name is a table in a FROM, JOIN,
+    /// or similar table clause. Currently, this is used only to support unquoted hyphenated identifiers
+    /// in this context on BigQuery.
+    pub fn parse_object_name(&mut self, in_table_clause: bool) -> Result<ObjectName, ParserError> {
+        self.parse_object_name_inner(in_table_clause, false)
+    }
+
+    /// Parse a possibly qualified, possibly quoted identifier, e.g.
+    /// `foo` or `myschema."table"
+    ///
+    /// The `in_table_clause` parameter indicates whether the object name is a table in a FROM, JOIN,
+    /// or similar table clause. Currently, this is used only to support unquoted hyphenated identifiers
+    /// in this context on BigQuery.
+    ///
+    /// The `allow_wildcards` parameter indicates whether to allow for wildcards in the object name
     /// e.g. *, *.*, `foo`.*, or "foo"."bar"
-    fn parse_object_name_with_wildcards(
+    fn parse_object_name_inner(
         &mut self,
         in_table_clause: bool,
         allow_wildcards: bool,
     ) -> Result<ObjectName, ParserError> {
-        let mut idents = vec![];
-
+        let mut parts = vec![];
         if dialect_of!(self is BigQueryDialect) && in_table_clause {
             loop {
                 let (ident, end_with_period) = self.parse_unquoted_hyphenated_identifier()?;
-                idents.push(ident);
+                parts.push(ObjectNamePart::Identifier(ident));
                 if !self.consume_token(&Token::Period) && !end_with_period {
                     break;
                 }
             }
         } else {
             loop {
-                let ident = if allow_wildcards && self.peek_token().token == Token::Mul {
+                if allow_wildcards && self.peek_token().token == Token::Mul {
                     let span = self.next_token().span;
-                    Ident {
+                    parts.push(ObjectNamePart::Identifier(Ident {
                         value: Token::Mul.to_string(),
                         quote_style: None,
                         span,
+                    }));
+                } else if let Some(func_part) =
+                    self.maybe_parse(|parser| parser.parse_object_name_function_part())?
```

Comment:
wondering can we instead extende the `else` clause to do something like this?
```rust
else {
    let ident = self.parse_identifier()?;
    let part = if self.dialect.is_identifier_generating_function_name(&ident) {
          self.expect_token(&Token::LParen)?;
          let args = ...
          self.expect_token(&Token::RParen)?;
         ObjectNamePartFunction { name:ident, args }
    } else {
          ObjectNamePart::Identifier(ident));
    }
}
```
thinking that way we skip this extra branch, and the extra error scenario in the function part parsing